# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

## 读取数据

读取titles.csv和credits.csv，将他们转化为DataFram，并分别复制给变量original_title_data和original_credits_data。

In [1]:
import pandas as pd

In [2]:
original_title_data = pd.read_csv('titles.csv')
original_credits_data = pd.read_csv('credits.csv')

In [3]:
original_title_data.head()

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
0,ts300399,Five Came Back: The Reference Films,SHOW,This collection includes 12 World War II-era p...,1945,TV-MA,51,['documentation'],['US'],1.0,NaN,NaN,NaN,0.600,NaN
1,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976,R,114,"['drama', 'crime']",['US'],NaN,tt0075314,8.2,808582.0,40.965,8.179
2,tm154986,Deliverance,MOVIE,Intent on seeing the Cahulawassee River before...,1972,R,109,"['drama', 'action', 'thriller', 'european']",['US'],NaN,tt0068473,7.7,107673.0,10.010,7.300
3,tm127384,Monty Python and the Holy Grail,MOVIE,"King Arthur, accompanied by his squire, recrui...",1975,PG,91,"['fantasy', 'action', 'comedy']",['GB'],NaN,tt0071853,8.2,534486.0,15.461,7.811
4,tm120801,The Dirty Dozen,MOVIE,12 American military prisoners in World War II...,1967,NaN,150,"['war', 'action']","['GB', 'US']",NaN,tt0061578,7.7,72662.0,20.398,7.600


In [4]:
original_credits_data.head()

,person_id,id,name,character,role
0,3748,tm84618,Robert De Niro,Travis Bickle,ACTOR
1,14658,tm84618,Jodie Foster,Iris Steensma,ACTOR
2,7064,tm84618,Albert Brooks,Tom,ACTOR
3,3739,tm84618,Harvey Keitel,Matthew 'Sport' Higgins,ACTOR
4,48933,tm84618,Cybill Shepherd,Betsy,ACTOR


## 数据评估和清洗

数据评估主要分为结构和内容两个方面。数据结构即整洁度评估标准为：是否符合每行是一个观察值，每列是一个变量，每个单元格是一个值的标准。数据内容即干净度评估标准为：是否有缺失数据、重复数据、指代不一致数据、无效和错误数据。

为了区分原始数据和清洗后的数据，需要创建新的变量cleaned_title_data和cleaned_credits_data，并将original_title_data和original_credits_data复制的副本赋值给它们。

In [5]:
cleaned_title_data = original_title_data.copy()
cleaned_credits_data = original_credits_data.copy()

In [6]:
cleaned_title_data.head()

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
0,ts300399,Five Came Back: The Reference Films,SHOW,This collection includes 12 World War II-era p...,1945,TV-MA,51,['documentation'],['US'],1.0,NaN,NaN,NaN,0.600,NaN
1,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976,R,114,"['drama', 'crime']",['US'],NaN,tt0075314,8.2,808582.0,40.965,8.179
2,tm154986,Deliverance,MOVIE,Intent on seeing the Cahulawassee River before...,1972,R,109,"['drama', 'action', 'thriller', 'european']",['US'],NaN,tt0068473,7.7,107673.0,10.010,7.300
3,tm127384,Monty Python and the Holy Grail,MOVIE,"King Arthur, accompanied by his squire, recrui...",1975,PG,91,"['fantasy', 'action', 'comedy']",['GB'],NaN,tt0071853,8.2,534486.0,15.461,7.811
4,tm120801,The Dirty Dozen,MOVIE,12 American military prisoners in World War II...,1967,NaN,150,"['war', 'action']","['GB', 'US']",NaN,tt0061578,7.7,72662.0,20.398,7.600


In [7]:
cleaned_credits_data.head()

,person_id,id,name,character,role
0,3748,tm84618,Robert De Niro,Travis Bickle,ACTOR
1,14658,tm84618,Jodie Foster,Iris Steensma,ACTOR
2,7064,tm84618,Albert Brooks,Tom,ACTOR
3,3739,tm84618,Harvey Keitel,Matthew 'Sport' Higgins,ACTOR
4,48933,tm84618,Cybill Shepherd,Betsy,ACTOR


### 数据整洁度

从cleaned_title_data中抽样10行数据观察。

In [8]:
cleaned_title_data.sample(10)

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
3466,ts225430,Mystic Pop-up Bar,SHOW,At this mysterious late-night food cart run by...,2020,NaN,70,"['drama', 'scifi', 'comedy', 'fantasy']",['KR'],1.0,tt12246190,8.0,2073.0,7.735,8.286
4764,ts320073,Attack on Pearl Harbor: Minute by Minute,SHOW,Covers the 1941 Japanese military strike on a ...,2021,TV-14,45,"['documentation', 'history']",['GB'],1.0,tt16348586,7.5,68.0,3.377,NaN
2568,ts74765,Daughters of Destiny,SHOW,Five girls from India's most impoverished fami...,2017,NaN,61,['documentation'],['IN'],1.0,tt1877895,8.6,922.0,2.962,7.700
172,tm189171,Wild Tango,MOVIE,A biopic based on the life of one of the pione...,1993,R,124,"['drama', 'music', 'romance']","['ES', 'AR']",NaN,tt0108291,6.7,1584.0,5.092,6.100
3043,ts216746,The Secret of Skinwalker Ranch,SHOW,A team of experts and scientists undertakes ex...,2020,TV-PG,42,"['horror', 'reality', 'thriller', 'documentati...",['US'],3.0,tt10589968,6.0,2625.0,30.911,8.500
4889,tm1075215,Aftermath,MOVIE,"Desperate to save their marriage, a young coup...",2021,NaN,114,"['horror', 'thriller', 'drama']",['US'],NaN,tt10691162,5.3,15646.0,46.058,6.400
1301,ts43209,The Cuba Libre Story,SHOW,"Recounts the tumultuous history of Cuba, a nat...",2015,TV-14,45,"['documentation', 'history', 'european']",['DE'],2.0,tt5319838,7.6,1649.0,3.412,6.800
4308,tm863611,Kappela,MOVIE,"Set in the village areas of Wayanad, Malappura...",2020,NaN,113,"['romance', 'comedy']",['IN'],NaN,tt11230868,7.5,3289.0,NaN,NaN
4772,ts102086,Behind Her Eyes,SHOW,"It follows Louise, a single mom with a son and...",2021,TV-MA,50,"['thriller', 'drama']",['GB'],1.0,tt9698442,7.2,59293.0,29.851,7.600
3911,tm823302,Alice Júnior,MOVIE,"This is the story of Alice Junior, a transgend...",2019,PG-13,90,"['romance', 'comedy', 'drama', 'animation']",['BR'],NaN,tt8217188,6.8,541.0,2.977,6.800


整洁数据需要符合每行是一个观察值，每列是一个变量，每个单元格是一个值的标准，从以上抽样来看，genres列和production_countries里的均是列表，存在一个单元格里有多个值的情况，因此需要将这些行进行拆分。

从genres提取出任意一个值分析。

In [9]:
cleaned_title_data['genres'][5579]

"['comedy', 'romance']"

发现这个值是字符串而不是字符串列表，无法使用explode进行拆分，所以需要先用eval函数将其转化成表达式，即字符串列表形式，才能进行拆分。

In [10]:
cleaned_title_data['genres'] = cleaned_title_data['genres'].apply(lambda x:eval(x))

验证genres里的字符串是否已转化为字符串列表：

In [11]:
cleaned_title_data['genres'][5579]

['comedy', 'romance']

确认转化为字符串列表后，机可对其进行拆分。

In [12]:
cleaned_title_data = cleaned_title_data.explode('genres')
cleaned_title_data

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
0,ts300399,Five Came Back: The Reference Films,SHOW,This collection includes 12 World War II-era p...,1945,TV-MA,51,documentation,['US'],1.0,NaN,NaN,NaN,0.600,NaN
1,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976,R,114,drama,['US'],NaN,tt0075314,8.2,808582.0,40.965,8.179
1,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976,R,114,crime,['US'],NaN,tt0075314,8.2,808582.0,40.965,8.179
2,tm154986,Deliverance,MOVIE,Intent on seeing the Cahulawassee River before...,1972,R,109,drama,['US'],NaN,tt0068473,7.7,107673.0,10.010,7.300
2,tm154986,Deliverance,MOVIE,Intent on seeing the Cahulawassee River before...,1972,R,109,action,['US'],NaN,tt0068473,7.7,107673.0,10.010,7.300
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5847,tm1059008,Lokillo,MOVIE,A controversial TV host and comedian who has b...,2021,NaN,90,comedy,['CO'],NaN,tt14585902,3.8,68.0,26.005,6.300
5848,tm1035612,Dad Stop Embarrassing Me - The Afterparty,MOVIE,"Jamie Foxx, David Alan Grier and more from the...",2021,PG-13,37,NaN,['US'],NaN,NaN,NaN,NaN,1.296,10.000
5849,ts271048,Mighty Little Bheem: Kite Festival,SHOW,"With winter behind them, Bheem and his townspe...",2021,NaN,7,family,[],1.0,tt13711094,7.8,18.0,2.289,10.000
5849,ts271048,Mighty Little Bheem: Kite Festival,SHOW,"With winter behind them, Bheem and his townspe...",2021,NaN,7,animation,[],1.0,tt13711094,7.8,18.0,2.289,10.000


处理production_countries列也是同样的方式，先提取其中任意一个值观察。

In [13]:
cleaned_title_data['production_countries'][5848]

"['US']"

该值为字符串而不是字符串列表，使用eval函数将其转换为列表。

In [14]:
cleaned_title_data['production_countries'] = cleaned_title_data['production_countries'].apply(lambda x:eval(x))

In [15]:
cleaned_title_data['production_countries'][5848]

['US']

确认转化为字符串列表后，使用explode将其拆分。

In [16]:
cleaned_title_data = cleaned_title_data.explode('production_countries')
cleaned_title_data.sample(10)

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
155,tm10811,Heavy,MOVIE,Victor is a cook who works in a greasy bar/res...,1995,R,105,drama,US,NaN,tt0113280,6.8,5047.0,3.308,6.800
1760,ts81208,Charmed,SHOW,Set in the fictional college town of Hilltowne...,2018,TV-14,40,drama,US,4.0,tt6394324,4.6,16533.0,92.694,7.427
3106,ts221200,The Haunting of Bly Manor,SHOW,"After an au pair’s tragic death, Henry Wingrav...",2020,TV-MA,55,thriller,US,1.0,tt10970552,7.3,99047.0,29.636,7.800
1386,tm235405,Hyena Road,MOVIE,"Three different men, three different worlds, t...",2015,R,120,war,CA,NaN,tt4034452,6.5,7839.0,9.363,6.700
1103,tm176553,Doctor Bello,MOVIE,Cancer specialist Dr. Michael Durant forms an ...,2013,NaN,95,drama,US,NaN,tt2723358,4.8,54.0,1.802,6.000
4824,tm471968,Passing,MOVIE,"In 1920s New York City, a Black woman finds he...",2021,PG-13,98,thriller,GB,NaN,tt8893974,6.7,13803.0,29.687,6.451
1809,tm244197,Replicas,MOVIE,A scientist becomes obsessed with returning hi...,2018,PG-13,94,scifi,GB,NaN,tt4154916,5.5,37340.0,27.644,6.100
202,tm256669,Gumrah,MOVIE,"Roshni Chadha makes her living, singing in var...",1993,NaN,149,crime,IN,NaN,tt0107060,6.1,795.0,1.763,6.700
3056,ts81007,After Life,SHOW,Tony had a perfect life. But after his wife Li...,2019,TV-MA,28,comedy,GB,3.0,tt8398600,8.4,129243.0,19.931,7.843
3048,ts87248,In the Dark,SHOW,Murphy is a flawed and irreverent woman who ju...,2019,TV-14,41,action,US,4.0,tt7772602,7.5,11409.0,39.136,7.900


这样数据cleaned_title_data就符合每行是一个观察值，每列是一个变量，每个单元格是一个值的要求，cleaned_title_data数据的整洁度清理完毕。

再从cleaned_credits_data中随机抽取10行数据观察。

In [17]:
cleaned_credits_data.sample(10)

,person_id,id,name,character,role
41530,1905339,tm350971,Iván Chávez,Jose,ACTOR
39402,200005,tm372054,Eva Martín,Natalia,ACTOR
26648,170550,tm271974,Ali Samadi Ahadi,NaN,DIRECTOR
58360,2252,tm885669,Kurt Russell,Santa Claus,ACTOR
45313,6210,tm427335,Reese Alexander,ADF Agent,ACTOR
61894,1057941,tm460411,Ahmet Olgun Sünear,NaN,ACTOR
9241,311872,ts27459,Kou Zhenhai,Doctor,ACTOR
14664,69484,tm182125,Deepti Naval,Rahila Qureshi,ACTOR
23263,6671,tm134516,Woon Young Park,Thunder Fist,ACTOR
74645,1219987,ts307981,Rhys Slack,Wyatt,ACTOR


cleaned_credits_data符合每行是一个观察值，每列是一个变量，每个单元格是一个值的整洁数据标准。

### 数据干净度

处理数据干净度主要是处理缺失数据、重复数据、不一致数据、无效和错误数据等方面的问题。

#### 数据cleaned_title_data内容的评估和清洗。

先使用info方法对数据内容有个整体了解。

In [18]:
cleaned_title_data.info()

<class 'pandas.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  str    
 1   title                 17817 non-null  str    
 2   type                  17818 non-null  str    
 3   description           17790 non-null  str    
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  str    
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  str    
 8   production_countries  17439 non-null  str    
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  str    
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), str(8)
memo

从输出的结果看，cleaned_title_data数据共有17818行数据，从非空值数量可以看出，title、description、age_certification、genres、production_countries、seasons、imdb_id、imdb_score、imdb_votes、tmdb_popularity、tmdb_score列存在空缺数据。release_year表示作品发布年份，数据类型应是日期而不是整数，需要转换数据类型。这些问题将在后续清洗步骤处理。

##### 数据类型转换

将release_year的数据类型转换为日期格式。

In [19]:
cleaned_title_data['release_year'] = pd.to_datetime(cleaned_title_data['release_year'], format='%Y')
cleaned_title_data['release_year']

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[us]

##### 处理空缺数据

由于数据分析的目的是整体不同流派作品的各演员IMDB平均评分，进而挖掘出各流派中评分最高的演员，因此在以上存在孔雀数据的变量中，只有genres和imdb_score是对分析结果产生影响的变量，其余变量存在缺失值对分析结果不产生影响，无需处理。

先提取genres列为空缺值的行观察。

In [20]:
cleaned_title_data[cleaned_title_data['genres'].isnull()]

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
212,tm255589,One Last Shot,MOVIE,"In this low-budget short film, two best buddie...",1998-01-01,NaN,30,NaN,NaN,NaN,NaN,NaN,NaN,1.890,5.2
619,tm341561,Like Twenty Impossibles,MOVIE,Occupied Palestine: A serene landscape now poc...,2003-01-01,NaN,16,NaN,PS,NaN,NaN,NaN,NaN,0.812,6.5
632,ts86241,Le Robe De Mariage Des Cieux,SHOW,It was with much difficulty that Ai Qing was a...,2004-01-01,TV-MA,63,NaN,NaN,1.0,NaN,NaN,NaN,0.600,NaN
636,tm404676,To and from New York,MOVIE,"While covering a story in New York City, a Sea...",2006-01-01,NaN,82,NaN,US,NaN,NaN,NaN,NaN,1.401,5.8
637,tm89054,Osuofia in London 2,MOVIE,Osuofia return to his Nigerian village with a ...,2004-01-01,NaN,72,NaN,XX,NaN,NaN,NaN,NaN,1.091,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5799,tm1040959,The Circle: The Afterparty,MOVIE,Stars of The Circle drop by to discuss Season ...,2021-01-01,NaN,35,NaN,US,NaN,NaN,NaN,NaN,1.882,10.0
5802,ts302434,Plastic Cup Boyz: Laughing My Mask Off!,SHOW,Comedy collective The Plastic Cup Boyz pour ou...,2021-01-01,NaN,33,NaN,NaN,1.0,NaN,NaN,NaN,0.683,NaN
5840,tm1216735,Sun of the Soil,MOVIE,"In 14th-century Mali, an ambitious young royal...",2022-01-01,NaN,26,NaN,NaN,NaN,NaN,NaN,NaN,1.179,7.0
5844,tm1074617,Bling Empire - The Afterparty,MOVIE,"The stars of ""Bling Empire"" discuss the show's...",2021-01-01,NaN,35,NaN,US,NaN,NaN,NaN,NaN,NaN,NaN


由于缺失数据分析所需的核心数据genres，因此需要将这些genres缺失的行删除，并检查删除后genres为空缺值的行个数。

In [21]:
cleaned_title_data = cleaned_title_data.dropna(subset=['genres'])
cleaned_title_data['genres'].isnull().sum()

np.int64(0)

将imdb_score为空值的行提取出来观察。

In [22]:
cleaned_title_data[cleaned_title_data['imdb_score'].isnull()]

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
0,ts300399,Five Came Back: The Reference Films,SHOW,This collection includes 12 World War II-era p...,1945-01-01,TV-MA,51,documentation,US,1.0,NaN,NaN,NaN,0.600,NaN
75,tm132164,Bill Hicks: Sane Man,MOVIE,Sane Man was filmed before Bill recorded ‘Dang...,1989-01-01,R,80,comedy,US,NaN,NaN,NaN,NaN,3.377,7.5
145,ts251477,My First Errand,SHOW,“Hajimete no Otsukai” (First Errand) is a Japa...,1991-01-01,TV-G,18,documentation,JP,12.0,NaN,NaN,NaN,7.730,7.8
145,ts251477,My First Errand,SHOW,“Hajimete no Otsukai” (First Errand) is a Japa...,1991-01-01,TV-G,18,family,JP,12.0,NaN,NaN,NaN,7.730,7.8
145,ts251477,My First Errand,SHOW,“Hajimete no Otsukai” (First Errand) is a Japa...,1991-01-01,TV-G,18,reality,JP,12.0,NaN,NaN,NaN,7.730,7.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5790,tm1094060,My Heroes Were Cowboys,MOVIE,Robin Wiltshire's painful childhood was rescue...,2021-01-01,PG,23,documentation,US,NaN,tt15084326,NaN,NaN,3.145,7.7
5791,tm1047429,Alan Saldaña: Locked Up,MOVIE,"Mexican comedian Alan Saldaña is back, poking ...",2021-01-01,NaN,49,comedy,NaN,NaN,NaN,NaN,NaN,6.670,6.0
5810,tm1225897,Social Man,MOVIE,Two competitive social media Influencers go he...,2021-01-01,NaN,96,comedy,NaN,NaN,tt20198164,NaN,NaN,NaN,NaN
5810,tm1225897,Social Man,MOVIE,Two competitive social media Influencers go he...,2021-01-01,NaN,96,drama,NaN,NaN,tt20198164,NaN,NaN,NaN,NaN


由于缺失数据分析所需的核心数据imdb_score，因此需要将imdb_score空缺的行删除，并检查删除后imdb_score为空缺的行数量。

In [23]:
cleaned_title_data = cleaned_title_data.dropna(subset=['imdb_score'])
cleaned_title_data['imdb_score'].isnull().sum()

np.int64(0)

##### 处理重复数据

从每个变量的含义来看，cleaned_title_data不应该存在完全相同的两行，因此检查是否存在重复数据。

In [24]:
cleaned_title_data[cleaned_title_data.duplicated()]

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score


按条件筛选出的重复行为空，因此不存在各变量完全相同的行。

##### 处理不一致数据

不一致数据指的是变量中存在不同值述指代同一目标的情况，根据各变量含义来看，genres和production_countries（出品国家）列可能存在这种情况，因此使用value_counts方法查看不同值的种类，以确定是否存在不一致数据。

In [25]:
cleaned_title_data['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

从输出结果来看，genres列不存在不一致数据。

In [26]:
cleaned_title_data['production_countries'].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
NP       1
LK       1
GT       1
AF       1
FO       1
Name: count, Length: 108, dtype: int64

输出结果中，由于production_countries的值种类太多，中间省略了一部分显示。为了完整显示，需要将display.max_rows设置为None，即取消显示上限，另外目前只是当前调用value_counts才需要全部显示，因此可以结合option_context临时更改显示上限。

In [27]:
with pd.option_context('display.max_rows', None):
    print(cleaned_title_data['production_countries'].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
ES          637
KR          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
HK          102
NL          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
AT           28
SA           28
IS           28
NZ           27
LU           27
PE           26
RO           25
QA           24
CZ           22
JO           19
HU           18
FI           18
MA           15
UY           15
PT           14
KW           10
KH           10
PK            9
PR 

从输出结果发现里面的值都是各国家两位字母代码，唯独Lebanon这个值是用英文单词，Lebanon的国家代码是LB，出现了70次，说明此处存在不一致，因此需要将Lebanon和LB统一成LB，即将Lebanon替换成LB，并在替换后检查Lebanon值的个数。

In [28]:
cleaned_title_data['production_countries'] = cleaned_title_data['production_countries'].replace('Lebanon', 'LB')
print(len(cleaned_title_data[cleaned_title_data['production_countries'] == 'Lebanon']))

0


检查结果为值为Lebanon的个数为0，确认已将production_countries列中的不一致数据清洗。

##### 无效和错误数据

使用describe方法查看cleaned_title_data数值变量的描述统计信息。

In [29]:
cleaned_title_data.describe()

,release_year,runtime,seasons,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
count,16970,16970.000000,5954.000000,16970.000000,1.694100e+04,16842.000000,16515.000000
mean,2015-11-14 22:42:51.974072,80.912552,2.455492,6.514207,3.281655e+04,29.396307,6.846933
min,1954-01-01 00:00:00,0.000000,1.000000,1.500000,5.000000e+00,0.600000,1.000000
25%,2015-01-01 00:00:00,45.000000,1.000000,5.800000,7.800000e+02,4.070000,6.200000
50%,2018-01-01 00:00:00,90.000000,2.000000,6.600000,3.508000e+03,10.195000,6.900000
75%,2020-01-01 00:00:00,107.000000,3.000000,7.300000,1.697800e+04,23.639000,7.500000
max,2022-01-01 00:00:00,225.000000,42.000000,9.500000,2.294231e+06,2274.044000,10.000000
std,NaN,39.596172,2.869428,1.131095,1.141492e+05,93.178235,1.078831


从输出结果来看，cleaned_title_data不存在脱离现实意义的数值。

#### 数据cleaned_credits_data内容的评估和清洗。

In [30]:
cleaned_credits_data.sample(5)

,person_id,id,name,character,role
72550,45431,tm1000619,Prabhas,Vikramaditya,ACTOR
6189,13515,tm58382,Desi Lydic,Cutie,ACTOR
29092,205664,tm350379,Mylène Farmer,Pauline,ACTOR
27839,4254,tm232797,Aneurin Barnard,Gibson,ACTOR
3494,63497,tm1822,Naoki Fuji,Mugger,ACTOR


先使用info方法对数据概况有所了解。

In [31]:
cleaned_credits_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   person_id  77801 non-null  int64
 1   id         77801 non-null  str  
 2   name       77801 non-null  str  
 3   character  68029 non-null  str  
 4   role       77801 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.0 MB


由输出结果来看，cleaned_credits_data一共有77801行数据。其中character列的非空值数量小于行数，因此该列存在非空值。另外person_id的含义是演职员ID，应为字符串，不应该是整数类型，需要转换数据类型。

##### 数据类型转换

In [32]:
cleaned_credits_data['person_id'] = cleaned_credits_data['person_id'].astype(str)
cleaned_credits_data['person_id']

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: str

##### 空缺数据

由以上评估结论，character变量存在空缺值，因此先提取出该列为空缺值的行。

In [33]:
cleaned_credits_data.query('character.isnull()')

,person_id,id,name,character,role
36,3308,tm84618,Martin Scorsese,NaN,DIRECTOR
59,17727,tm154986,John Boorman,NaN,DIRECTOR
106,11475,tm127384,Terry Jones,NaN,DIRECTOR
107,11473,tm127384,Terry Gilliam,NaN,DIRECTOR
162,1063,tm120801,Robert Aldrich,NaN,DIRECTOR
...,...,...,...,...,...
77776,2363022,tm1097142,Mohamed El-Arkan,NaN,ACTOR
77777,1827884,tm1097142,Mohamed Bakir,NaN,DIRECTOR
77783,678884,tm1014599,Segun Arinze,NaN,ACTOR
77789,1962840,tm1014599,Seyi Babatope,NaN,DIRECTOR


共提取出9772行character列为空缺值的观察值，由于character并非是影响分析结果的关键数据，因此可以不对该列的空缺值处理。

##### 重复数据

根据各变量的含义，cleaned_credits_data不应存在所有变量都相等的重复数据，因此检查可能存在的重复数据。

In [34]:
len(cleaned_credits_data[cleaned_credits_data.duplicated()])

0

##### 不一致数据

根据cleaned_credits_data各变量含义来看，只有role变量可能存在不一致数据，即不同值指代同一角色的情况，因此调用value_counts方法来查看role变量中各个值出现的次数。

In [35]:
cleaned_credits_data['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

根据输出结果来看，role变量中只有ACTOR和DIRECTOR两种值，分别指代演员和导演，不存在指代不一致的情况。另外由于该列只有两种值，所以可以将role变量的数据类型转换为category，这样可以节省内存空间。

In [36]:
cleaned_credits_data['role'] = cleaned_credits_data['role'].astype('category')
cleaned_credits_data['role']

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, str): ['ACTOR', 'DIRECTOR']

##### 无效和错误数据

由于cleaned_credits_data不存在数值类型的数据，因此不需要对其调用describe方法检查。

## 数据整理

In [37]:
cleaned_title_data.sample(10)

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score
4136,tm849937,Jungle Beat: The Movie,MOVIE,The Jungle Beat animals think it’s the best th...,2020-01-01,G,84,scifi,MU,NaN,tt12139700,5.4,1029.0,21.935,7.0
1565,tm236705,Ricardo O'Farrill: Abrazo Genial,MOVIE,"Ricardo O'Farrill brings his sharp, observatio...",2016-01-01,NaN,92,comedy,MX,NaN,tt5547496,5.5,117.0,1.481,5.2
2056,ts83232,The Mire,SHOW,The town is shocked with a brutal double murde...,2018-01-01,TV-MA,54,thriller,PL,2.0,tt8855592,7.0,4355.0,13.039,6.8
89,tm52274,The Little Wars,MOVIE,Beirut resident Soraya is drawn to two men: da...,1982-01-01,NaN,108,drama,LB,NaN,tt0084492,6.3,152.0,1.956,5.5
1664,tm211571,Niyazi Gül Dörtnala,MOVIE,"Niyazi, a veterinarian, is on the lookout for ...",2015-01-01,PG,106,comedy,TR,NaN,tt5016660,4.5,3170.0,2.395,4.2
3860,ts218890,Drifting Dragons,SHOW,"Dragons, the rulers of the sky. To many people...",2020-01-01,TV-14,24,scifi,JP,1.0,tt11418344,6.0,591.0,6.905,7.7
1413,tm246175,The 101-Year-Old Man Who Skipped Out on the Bi...,MOVIE,Drinking the tasty Folk Soda puts a spring in ...,2016-01-01,NaN,108,european,SE,NaN,tt4940370,6.3,6023.0,7.299,6.4
1078,ts32752,Two Fathers,SHOW,"Seven years ago, two men were notified that th...",2013-01-01,TV-14,44,drama,TW,1.0,tt6707142,8.0,59.0,1.762,7.0
185,tm110766,Chronicle of a Disappearance,MOVIE,Chronicle of a Disappearance unfolds in a seri...,1996-01-01,NaN,88,drama,DE,NaN,tt0115895,6.9,733.0,2.099,6.7
4422,tm912415,Urzila Carlson: Overqualified Loser,MOVIE,Stand-up comedian Urzila Carlson keeps the cro...,2020-01-01,NaN,62,comedy,NaN,NaN,tt12572032,6.3,366.0,2.950,4.9


In [38]:
cleaned_credits_data.sample(10)

,person_id,id,name,character,role
74366,51682,tm1120937,Nathan Frankowski,NaN,DIRECTOR
16999,5772,tm171042,Abigail Breslin,Casey Welson,ACTOR
57144,115686,ts84639,Soma Saito,Keisaku Asano,ACTOR
52788,1485898,ts242244,George Cockle,Daniel Zakovic (voice),ACTOR
12274,69034,tm85369,Paresh Rawal,Hemu Patel,ACTOR
61144,1294153,tm848444,Yousef Aldakheel,Satam,ACTOR
2412,9309,tm192405,Una Damon,Head Nurse,ACTOR
47784,884318,tm332071,Monique McKellop,LYFT Driver (uncredited),ACTOR
6298,79529,tm35463,Jimmy Burke,FBI Waiter #4,ACTOR
51375,327760,tm460829,Seishuu Uragami,Sora (voice),ACTOR


由于该数据分析的目的是整理不同流派的各演员参演作品的IMDB平均评分，从而挖掘出各流派中评分最高的演员，而流派（genres）、IMDB评分（imdb_score）、演职员ID（person_id）分别属于不同的DataFrame,如果想要得到不同流派的各演员参演作品的IMDB平均评分，就需要将cleaned_title_data和cleaned_credits_data两个DataFrame通过共同的键id进行合并，我们将合并后的DataFrame取名为title_and_credits。

In [39]:
title_and_credits = pd.merge(cleaned_title_data, cleaned_credits_data, on ='id')
title_and_credits.head()

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score,person_id,name,character,role
0,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976-01-01,R,114,drama,US,NaN,tt0075314,8.2,808582.0,40.965,8.179,3748,Robert De Niro,Travis Bickle,ACTOR
1,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976-01-01,R,114,drama,US,NaN,tt0075314,8.2,808582.0,40.965,8.179,14658,Jodie Foster,Iris Steensma,ACTOR
2,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976-01-01,R,114,drama,US,NaN,tt0075314,8.2,808582.0,40.965,8.179,7064,Albert Brooks,Tom,ACTOR
3,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976-01-01,R,114,drama,US,NaN,tt0075314,8.2,808582.0,40.965,8.179,3739,Harvey Keitel,Matthew 'Sport' Higgins,ACTOR
4,tm84618,Taxi Driver,MOVIE,A mentally unstable Vietnam War veteran works ...,1976-01-01,R,114,drama,US,NaN,tt0075314,8.2,808582.0,40.965,8.179,48933,Cybill Shepherd,Betsy,ACTOR


在role变量中还含有DIRECTOR（导演）的值，由于是要整理各流派影视作品中各演员的IMDB评分，只包含演员并不包含导演，因此需要将role变量的值为DIRECTOR（导演）的列删除，并检查删除后数据title_and_credits中role的值为DIRECTOR的行数。

In [40]:
title_and_credits = title_and_credits.query('role != "DIRECTOR"')
title_and_credits.query('role == "DIRECTOR"')

,id,title,type,description,release_year,age_certification,runtime,genres,production_countries,seasons,imdb_id,imdb_score,imdb_votes,tmdb_popularity,tmdb_score,person_id,name,character,role


以上根据条件筛选出role值为DIRECTOR的DataFrame为空，说明已成功将role值为DIRECTOR的行删除。

接下来将title_and_credits按照流派（genres）和演员ID（person_id）为索引分组，提取imdb_score变量，计算出IMDB评分的平均值。

In [41]:
score_groupby_genres_and_person = title_and_credits.groupby(['genres','person_id'])['imdb_score'].mean()
score_groupby_genres_and_person

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

调用reset_index方法，重置索引，这样会得到一个规范的DataFrame。

In [44]:
score_groupby_genres_and_person = score_groupby_genres_and_person.reset_index()
score_groupby_genres_and_person

,genres,person_id,imdb_score
0,action,1000,6.866667
1,action,100007,7.000000
2,action,100013,6.400000
3,action,100019,6.500000
4,action,100020,6.500000
...,...,...,...
168876,western,993735,6.500000
168877,western,998673,7.300000
168878,western,998674,7.300000
168879,western,998675,7.300000


将重置索引得到的score_groupby_genres_and_person按照流派（genres）分组，提取imdb_score的最大值，得到各流派IMDB最大值的Series。

In [46]:
max_score_groupby_genres = score_groupby_genres_and_person.groupby('genres')['imdb_score'].max()
max_score_groupby_genres

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

将max_score_groupby_genres与score_groupby_genres_and_person以genres和imdb_score为键进行内连接，这样就能筛选出各流派影视作品中最高分的演员ID（person_id）。

In [49]:
max_score_groupby_genres_and_person = pd.merge(score_groupby_genres_and_person, max_score_groupby_genres, on=['genres', 'imdb_score'], how='inner')
max_score_groupby_genres_and_person

,genres,person_id,imdb_score
0,action,12790,9.3
1,action,1303,9.3
2,action,21033,9.3
3,action,336830,9.3
4,action,86591,9.3
...,...,...,...
131,war,826547,8.8
132,western,22311,8.9
133,western,28166,8.9
134,western,28180,8.9


这样得到了各流派影视作品中IMDB评分最高的演员ID，其中各个流派中的最高分演员可能不止一位。

接下来还需要知道各流派作品中评分最高的演员ID对应的名字，因此需要将max_score_groupby_genres_and_person和cleaned_credits_data进行连接，因为cleaned_credits_data还有其他列，而我们只需要用到person_id和name这两列，因此先提取这两列，再删除重复行。

In [50]:
person_and_name = cleaned_credits_data[['person_id','name']].drop_duplicates()
person_and_name

,person_id,name
0,3748,Robert De Niro
1,14658,Jodie Foster
2,7064,Albert Brooks
3,3739,Harvey Keitel
4,48933,Cybill Shepherd
...,...,...
77796,736339,Adelaida Buscato
77797,399499,Luz Stella Luengas
77798,373198,Inés Prieto
77799,378132,Isabel Gaona


将max_score_groupby_genres_and_person与person_and_name以person_id为键进行连接，得到包含各流派最高评分作品演员ID对应的演员姓名。

In [51]:
max_score_actor_goupby_genres = pd.merge(max_score_groupby_genres_and_person, person_and_name, on = 'person_id')
max_score_actor_goupby_genres

,genres,person_id,imdb_score,name
0,action,12790,9.3,Olivia Hack
1,action,1303,9.3,Jessie Flower
2,action,21033,9.3,Zach Tyler
3,action,336830,9.3,André Sogliuzzo
4,action,86591,9.3,Cricket Leigh
...,...,...,...,...
131,war,826547,8.8,Yuto Uemura
132,western,22311,8.9,Koichi Yamadera
133,western,28166,8.9,Megumi Hayashibara
134,western,28180,8.9,Unsho Ishizuka
